tc/core/execution_engine_api_rfc.h

```diff
@@ -0,0 +1,147 @@
+ */
+template <typename Backend>
+class ExecutionEngine {
+ public:
+  using BackendType = Backend;
+  using ExecutorType = typename BackendType::ExecutorType;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+
+  ExecutionEngine() = default;
+
+  /// Parse TC definitions provided as string, store parsed trees internally.
+  void define(const std::string& language);
+
+  /// Store the provided parsed trees internally.
+  void define(const std::vector<lang::TreeRef>& treeRefs);
+
+  /// For a given TC kernel name, compute the shapes of the output tensors
+  /// provided the shapes of the input TC tensors.  The caller can use the
+  /// computed shapes to allocate memory for outputs.
+  std::vector<TensorInfo> inferOutputTensorInfo(
+      const std::string& name,
+      const std::vector<const DLConstTensor*>& inTensorPtrs);
+
+  /// JIT-compile given the TC kernel name, the shapes of the input tensors and
+  /// the compilation options.  Must be overridden by a specific
+  /// ExecutionEngine, which also interprets the options as it sees fit.
+  /// \returns opaque handle of a compiled kernel.
+  size_t compile(
+      const std::string& name,
+      const std::vector<const DLConstTensor*>& inputs,
+      const MappingOptionsType& options);
+
+  /// Run a compiled TC kernel given its handle, on the given input tensors and
+  /// fill in the outputs.  All tensors must be allocated and have appropriate
+  /// shapes (inputs same as for copmilation, outputs same as returned by
+  /// inferOutputTensorInfo).
+  /// \returns The pair of (cpu,kernel) runtimes. If profile is set, the
+  /// kernel is profiled otherwise its duration is set to Duration::max().
+  ProfilingInfo run(
+      size_t handle,
+      const std::vector<const DLConstTensor*>& inputs,
+      const std::vector<const DLTensor*>& outputs,
+      bool profileKernel = false,
+      std::function<bool(const ExecutorType*)> pruningFunction =
+          [](const ExecutorType*) { return false; });
+
+  /// "Low-latency" execution mode in which we just propagate raw pointers to
+  /// data in GPU address space.
+  /// No tensor-related information can be checked so it is the user's
+  /// responsibility to ensure that shapes and strides match.
+  Duration uncheckedRun(
+      size_t handle,
+      const std::vector<const void*>& inputs,
+      const std::vector<void*>& outputs,
+      bool profileKernel = false);
+
+  /// Clear the compilation result for the given handle.
+  void clear(size_t handle);
+
+ private:
+  size_t emplaceExecutor(std::unique_ptr<ExecutorType>&& p);
+
+  size_t getHandle(
+      const std::string& name,
+      const std::vector<const DLConstTensor*>& inputs,
+      const MappingOptionsType& options);
+
+ private:
+  /// For thread-safety perform all cheap operations under lock.
+  std::mutex tcExecutorMutex_;
+
+  /// Parsed TC trees.
+  std::map<std::string, lang::TreeRef> tcNameMap_;
```

Comment:
This can be another class which stores all of the parsed TCs. Users don't have to use it, IMO, it is just a nice feature for some use-cases. In a basic use-case user can figure out caching theirselves, I think.